### PR TITLE
Adds retry support

### PR DESCRIPTION
**Pull request by tonyredondo**
This `PR` adds retry support of failing tests.

The way to activate it is by using the `agent.WithRetriesOnFail` option in the `Run` method, is the only modification needed. If the option is not set then the standard test run executes.

By default in golang when a test produces a panic, the whole test suite crash. Because we are trying to apply retries in failing test the options has two arguments:

```go
func WithRetriesOnFail(retriesCount int, exitOnError bool) Option {
	return func(agent *Agent) {
		agent.failRetriesCount = retriesCount
		agent.exitOnError = exitOnError
	}
}
```

- The retries count.
- An option to recover from panic on a test in order to retry it

image

### CODE REVIEW

File: runner/main.go [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+	}
+	return nil, errors.New("field can't be retrieved")
+}
+
+func (r *testRunner) getFqnOfTest(tFunc func(*testing.T)) string {
+	funcVal := runtime.FuncForPC(reflect.ValueOf(tFunc).Pointer())
+	return funcVal.Name()
+}
+
+func (r *testRunner) getFqnOfBenchmark(bFunc func(*testing.B)) string {
```
Comment (fermayo) [inline runner/main.go]:
```suggestion
func getFqnOfBenchmark(bFunc func(*testing.B)) string {
```

File: runner/main.go [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+	val := reflect.Indirect(reflect.ValueOf(r.m))
+	member := val.FieldByName(fieldName)
+	if member.IsValid() {
+		ptrToY := unsafe.Pointer(member.UnsafeAddr())
+		return ptrToY, nil
+	}
+	return nil, errors.New("field can't be retrieved")
+}
+
+func (r *testRunner) getFqnOfTest(tFunc func(*testing.T)) string {
```
Comment (fermayo) [inline runner/main.go]:
```suggestion
func getFqnOfTest(tFunc func(*testing.T)) string {
```

File: runner/main.go [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+				fqn:       fqn,
+			}
+			benchmarks = append(benchmarks, benchmark)
+		}
+		// Replace internal benchmark
+		*r.intBenchmarks = benchmarks
+	}
+}
+
+func (r *testRunner) getFieldPointerOfM(fieldName string) (unsafe.Pointer, error) {
```
Comment (fermayo) [inline runner/main.go]:
```suggestion
func getFieldPointerOfM(m *testing.M, fieldName string) (unsafe.Pointer, error) {
```

File: runner/main.go [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+	val := reflect.Indirect(reflect.ValueOf(t))
+	member := val.FieldByName(fieldName)
+	if member.IsValid() {
+		ptrToY := unsafe.Pointer(member.UnsafeAddr())
+		return ptrToY, nil
+	}
+	return nil, errors.New("field can't be retrieved")
+}
+
+func (r *testRunner) getTestResultsInfo(t *testing.T) *internalTestResult {
```
Comment (fermayo) [inline runner/main.go]:
```suggestion
func getTestResultsInfo(t *testing.T) *internalTestResult {
```

File: runner/main.go [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+	return r.exitCode
+}
+
+// Internal test processor, each test calls the processor in order to handle retries, process exiting and exitcodes
+func (r *testRunner) testProcessor(t *testing.T) {
+	t.Helper()
+	name := t.Name()
+	if item, ok := (*r.tests)[name]; ok {
+
+		//Sets the original test name
```
Comment (fermayo) [inline runner/main.go]:
```suggestion
		// Sets the original test name
```
Be careful with the formatting of comments

File: runner/main.go [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+		m:                m,
+		exitOnError:      exitOnError,
+		failRetriesCount: failRetriesCount,
+	}
+	runner.init()
+	return runner.Run()
+}
+
+// Gets the test name
+func GetTestName(name string) string {
```
Comment (fermayo) [inline runner/main.go]: Not a very descriptive function name. What about:
```suggestion
func GetOriginalTestName(name string) string {
```

File: runner/main.go [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+				*(*bool)(ptr) = false
+			}
+		}
+	} else {
+		t.FailNow()
+	}
+}
+
+// Initialize test runner
+func (r *testRunner) init() {
```
Comment (fermayo) [inline runner/main.go]: Nit picking: I'd move this method up before `testProcessor`, it's easer to read the file

File: runner/main.go [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+			(*r.tests)[fqn] = &testDescriptor{
+				test:   test,
+				fqn:    fqn,
+				ran:    0,
+				failed: false,
+			}
+			tests = append(tests, testing.InternalTest{
+				Name: fqn,
+				F:    r.testProcessor,
+			})
```
Comment (fermayo) [inline runner/main.go]: Instead of having a "parallel" data structure `r.tests` to keep metadata about each test, why not wrap each test into an object, that has a method to "run" it? (basically the `testProcessor` method). I think it will be cleaner and easier to read.

File: runner/main.go [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+// Runs the test suite
+func (r *testRunner) Run() int {
+	r.exitCode = 0
+	r.m.Run()
+	return r.exitCode
+}
+
+// Internal test processor, each test calls the processor in order to handle retries, process exiting and exitcodes
+func (r *testRunner) testProcessor(t *testing.T) {
+	t.Helper()
```
Comment (fermayo) [inline runner/main.go]: Why is this needed?

File: runner/main.go [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+			run++
+		}
+		if item.flaky {
+			fmt.Println("*** FLAKY", item.fqn)
+		}
+		if item.error && exitOnError {
+			panic(rc)
+		}
+		if item.error || item.failed {
+			r.exitCode = 1
```
Comment (fermayo) [inline runner/main.go]: Why do we need to track this? Can we just rely on `m.Run`?

File: runner/main.go (hunk 6 above)
Comment (tonyredondo) [reply runner/main.go]: ok

File: runner/main.go (hunk 9 above)
Comment (tonyredondo) [reply runner/main.go]: This is because I did not wanted to intercede in the normal stacktrace because I'm using the runner to execute the test. But maybe this doesn't help, I'll remove it.

File: runner/main.go (hunk 5 above)
Comment (tonyredondo) [reply runner/main.go]: ok

File: runner/main.go (hunk 10 above)
Comment (tonyredondo) [reply runner/main.go]: Because `m.Run` return 1 if a sub test fail, so in flaky tests with both fail and pass status, always will be returning 1, and I think we want to have a green build besides flakyness. That`s why I have to control the exit code of the runner.

File: runner/main.go (hunk 7 above)
Comment (tonyredondo) [reply runner/main.go]: ok, you're right

File: runner/main.go (hunk 3 above)
Comment (tonyredondo) [reply runner/main.go]: ok

File: runner/main.go (hunk 2 above)
Comment (tonyredondo) [reply runner/main.go]: ok

File: runner/main.go (hunk 1 above)
Comment (tonyredondo) [reply runner/main.go]: ok

File: runner/main.go (hunk 4 above)
Comment (tonyredondo) [reply runner/main.go]: ok

File: runner/main.go (hunk 10 above)
Comment (tonyredondo) [reply runner/main.go]: forget it, I found a way to change the value in `m.Run` so we don't need to track the exitCode anymore.

File: agent/agent.go [hunk 11]
```diff
@@ -107,12 +111,21 @@ func WithGitInfo(repository string, commitSha string, sourceRoot string) Option
 	}
 }
 
+func WithRetriesOnFail(retriesCount int, exitOnError bool) Option {
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply agent/agent.go]: Correct.

File: runner/main.go (hunk 8 above)
Comment (tonyredondo) [reply runner/main.go]: done

**Comment by tonyredondo** [comment]
@fermayo I changed a lot stuff here, please review it.

File: runner/main.go [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+			*(*bool)(ptr) = false
+			if parentPtr, err := getFieldPointerOfT(t, "parent"); err == nil {
+				parentTPointer := (**testing.T)(parentPtr)
+				if parentTPointer != nil && *parentTPointer != nil {
+					removeTestFailureFlag(*parentTPointer)
+				}
+			}
+		}
+	}
+}
```
Comment (fermayo) [inline runner/main.go]: As discussed offline, this will mark the whole suite as passing whenever there is a flaky test, even if there were tests that previously failed

File: runner/main.go [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+			innerTest = it
+			td.test.F(it)
+		})
+		innerTestInfo := getTestResultsInfo(innerTest)
+
+		if rc != nil {
+			if exitOnError {
+				panic(rc)
+			}
+			fmt.Println("PANIC RECOVER:", rc)
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply runner/main.go]: ok

File: runner/main.go [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+			break
+		}
+
+		if td.skipped || !td.failed || run > maxRetries {
+			break
+		}
+		run++
+	}
+	if td.flaky {
+		fmt.Println("*** FLAKY", t.Name())
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply runner/main.go]: ok

File: runner/main.go [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+			defer func() {
+				rc = recover()
+				if rc != nil {
+					it.FailNow()
+				}
+			}()
+			innerTest = it
+			td.test.F(it)
+		})
+		innerTestInfo := getTestResultsInfo(innerTest)
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply runner/main.go]: Yes, you're right. Replacing...

File: runner/main.go [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+			break
+		}
+		run++
+	}
+	if td.flaky {
+		fmt.Println("*** FLAKY", t.Name())
+	}
+	if td.error && exitOnError {
+		panic(rc)
+	}
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply runner/main.go]: It will. I've always recover from panic in the subtest for flaky test handling, so if the `exitOnError` flag is set, `rc` is the original panic info from the subtest.

File: runner/main.go (hunk 12 above)
Comment (tonyredondo) [reply runner/main.go]: right, I'm splitting this method in two, `setTestFailureFlag(t *testing.T, value bool)` and `getTestParent(t *testing.T) *testing.T`
